### PR TITLE
Support TRUNCATE from the MX worker nodes

### LINKED ISSUE #2344: Enable `TRUNCATE distributed/reference table` on MX workers
Currently, Citus errors out:

```SQL
TRUNCATE users_table;
ERROR:  operation is not allowed on this node
HINT:  Connect to the coordinator and run it again.
```

This is disabled because it is likely that a concurrent `SELECT/DML` on another node could create a distributed deadlock. The distributed deadlock could happen with the following steps:

```
1. A user connects to node-1 and executes `TRUNCATE table`;
2. `TRUNCATE ...` is executed on the `shard-1`
3. Another user connects to node-2 and executes `BEGIN; SELECT count(*) FROM table`
4. `BEGIN; SELECT ...` is executed on `shard-2`
5. `TRUNCATE ...` is blocked on the `shard-2` since the `SELECT ...` is still active
6. `SELECT ...` is blocked on the `shard-1` since the `TRUNCATE ...` is still active
7. Distributed deadlock is detected and one of the transactions is cancelled
```

### PR DESCRIPTION
**Pull request by onderkalaci**
DESCRIPTION: Enable TRUNCATE command support from MX worker nodes

Fixes #2344

This commit enables support for TRUNCATE on both distributed table and reference tables.
    
The basic idea is to acquire lock on the relation by sending the TRUNCATE command to all metadata worker nodes. We only skip sending the TRUNCATE command to the node that executes the command to prevent a self-distributed-deadlock.

After #2336 we should add an isolation test.

(Note that we've initially considered acquiring an explicit lock on the table via `LOCK TABLE ... `. However, this approach seems more straight-forward and generic enough to be applied in other cases as well.)

### CODE REVIEW

File: src/include/distributed/worker_transaction.h [hunk 1]
```diff
@@ -22,6 +22,7 @@
 typedef enum TargetWorkerSet
 {
 	WORKERS_WITH_METADATA,
+	WORKERS_WITH_METADATA_EXCLUDE_CURRENT_WORKER,
```
Comment (onderkalaci) [inline src/include/distributed/worker_transaction.h]: I'm open for alternative names for this enum. It seems too long now.

**Comment by mtuncer** [comment]
missing DESCRIPTION

File: src/include/distributed/worker_transaction.h (hunk 1 above)
Comment (mtuncer) [reply src/include/distributed/worker_transaction.h]: `OTHER_WORKERS_WITH_METADATA`  ?

File: src/backend/distributed/master/master_modify_multiple_shards.c [hunk 2]
```diff
@@ -196,6 +203,30 @@ master_modify_multiple_shards(PG_FUNCTION_ARGS)
...
 
+	/*
+	 * We should execute "TRUNCATE table_name;" on the other worker nodes before
+	 * executing the truncate commands on the shards. This is necessary to prevent
+	 * distributed deadlocks where a concurrent operation on the same table (or a
+	 * cascading table) is executed on the other nodes.
+	 *
+	 * Note that we should skip the current node to prevent a self-deadlock.
+	 */
+	if (ShouldSyncTableMetadata(relationId) && truncateOperation)
```
Comment (mtuncer) [inline src/backend/distributed/master/master_modify_multiple_shards.c]: minor, consider reversing arguments here. I think having truncateOperation check first makes it more readable

File: src/include/distributed/worker_transaction.h (hunk 1 above)
Comment (mtuncer) [reply src/include/distributed/worker_transaction.h]: I suggested `OTHER_WORKERS_WITH_METADATA` but it has not come here.

File: src/backend/distributed/transaction/worker_transaction.c [hunk 3]
```diff
@@ -117,7 +117,15 @@ SendBareCommandListToWorkers(TargetWorkerSet targetWorkerSet, List *commandList)
 		int nodePort = workerNode->workerPort;
 		int connectionFlags = FORCE_NEW_CONNECTION;
 
-		if (targetWorkerSet == WORKERS_WITH_METADATA && !workerNode->hasMetadata)
+		if ((targetWorkerSet == WORKERS_WITH_METADATA ||
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/worker_transaction.c]: Ops, this approach looks wrong :/  Workers may not know about the other workers. See #2320. 

I think we need to work on this again. We might consider the initial idea where any node issues something like `LOCK TABLE IF EXISTS x IN EXLUSIVE MODE ` to all other nodes.

And, we should never have something like `OTHER_WORKERS_WITH_METADATA `

**Pull request merged**